Sixth app - React Redux, use redux store to manage products

violentr opened the pull request:
Entry level of playing with React Redux
```
 - Create Redux store
 - Connect Redux store to the application

- Create actionTypes for Product and Products
- Create actions for Product and Products
- Create reducers for Product and Products
- Connect ProductContainer to use Redux store
- Make list of products available on componentDIdMount lifecycle

- Create actionTypes for Basket
- Create actions for Basket
- Create reducer for Basket
- Connect CartContainer to use Redux store (Basket)
- Make changes to reflect use of Redux store with CartContainer (Basket)
- Clean up CartContainer, make sure only Redux store used instead of component's state

- Remove not used code, whenever is possible to do so (Refactor)

violentr committed:
Add Redux libs, redux, redux-thank, react-redux

violentr committed:
Add Reducers related to the Product

violentr committed:
Add Redux store

violentr committed:
Add Api related info as contstant in constants

violentr committed:
Add Product and Products actions

violentr committed:
Add actionTypes for Product and Products

violentr committed:
Add Redux store to the App, integration

violentr committed:
Add lodash & qs libs

violentr committed:
Add lib redux-devtools

violentr committed:
Add redux-devtools, ability to use as React Component

violentr committed:
Add prepareData, make sure data is available on routes

violentr committed:
Add Component for checking state in Redux

Redux Devtools
https://github.com/reduxjs/redux-devtools/blob/master/docs/Walkthrough.md#manual-integration

violentr committed:
Add logic to show redux state on dif pathnames

violentr committed:
Add Products data fetched with help of Redux

violentr committed:
Add actions for Basket

violentr committed:
Add actionTypes for Basket

violentr committed:
Add Reducer for Basket, add_to_basket type

violentr committed:
Add Basket reducer to list of reducers

violentr committed:
Add connect Redux store to CartContainer

Redux store connected to the Component
fix code reflect this change, make
sure component's state not used
instead used Redux store

violentr committed:
Fix app name

index01d commented on Week-1/src/containers/CartContainer.js:
```diff
@@ -1,44 +1,39 @@
+import {connect} from 'react-redux'
+import {addProductToBasket} from '~/src/actions/Basket.js'
 
 class CartContainer extends Component {
```
CartContainer отсюда и далее уже не выглядит нужным компонентом, можно обернуть redux-connect'ом непосредственно презентационный компонент, который пользуется данными

index01d commented on Week-1/src/containers/CartContainer.js:
```diff
@@ -1,44 +1,39 @@
-    return(
+    return (
       <Fragment>
         <BasketProvider value={options}>
```
получается, что с началом использования react-redux старый реакт-контекст для этой задачи нам уже не нужен

index01d commented on Week-1/src/containers/CartContainer.js:
```diff
@@ -1,44 +1,39 @@
-  totalItems = () => {
-    return this.state.basket.length
-  }
+  totalItems = () => (
```
можно посчитать в statetToProps

index01d commented on Week-1/src/containers/CartContainer.js:
```diff
@@ -1,44 +1,39 @@
+  )
 
-  totalCost = (products) => {
+  totalCost = () => {
```
можно посчитать в stateToProps

index01d commented on Week-1/src/containers/CartContainer.js:
```diff
@@ -1,44 +1,39 @@
   }
 
   showBasket = () =>{
     this.props.history.push(checkoutPath())
```
можно заменить на `<Link to={checkoutPath()}>` в месте использования

index01d commented on Week-1/src/containers/ProductContainer.js:
```diff
@@ -1,41 +1,30 @@
+import {connect} from 'react-redux'
+import {fetchProducts} from '~/src/actions/Products.js'
 
 class ProductContainer extends Component {
```
аналогично это старое поведение можно также устранить и посчитать нужные значения в stateToProps

violentr committed:
Remove Container component CartContainer

No use any more as we are using Redux store
no need to have Cart Context, state read
from Redux store

violentr committed:
Remove BasketContext no need, Redux used instead

violentr committed:
Add Fix Basket to use Redux store instead of context

violentr committed:
Add Fix CheckoutPage use Redux store instead of context

violentr committed:
Remove Product context with container, Redux store used

No need to have Container component and its context
Redux store use instead

violentr committed:
Add Fix code to use Redux store not container context

violentr committed:
Better naming for initialStore

index01d commented on Week-1/src/components/Layout.js:
```diff
@@ -27,7 +25,7 @@ const Layout = () =>(
     <Switch>
```
вот эта часть кстати (`Switch...`) должна была остаться в App.js и сюда передаваться children'aми, ведь это часть настройки роутера а не layout'a

index01d commented on Week-1/src/components/Product/AddToBasket.js:
```diff
@@ -1,27 +1,34 @@
-import React, { Fragment } from 'react'
-import { BasketConsumer } from '~/src/context/BasketContext.js'
-import PropTypes from 'prop-types'
+import React, {Fragment, Component} from 'react'
```
небольшой лайфхак: чаще всего работать с презентационными компонентами и с кодом connect() приходится по-отдельности, поэтому чтобы избежать лишних соблазнов лучше держать по отдельности
-AddToBasket
--index.js // connect() и его составляющие (stateToProps, actionsToProps, mergeProps)
--AddToBasket.js // презентационный компонент

index01d commented on Week-1/src/components/Product/AddToBasket.js:
```diff
@@ -1,27 +1,34 @@
+}
+
+const mapStateToProps = (state) => (
+  state
```
вот так, к слову о лишних соблазнах, делать не нужно :) передавать в свойствах нужно только реально используемые данные. если из состояния ничего не нужно, то вместо stateToProps можно передать функции connect null, например: `connect(null, actionsToProps)(...)`

index01d commented on Week-1/src/components/Product/Basket.js:
```diff
@@ -1,14 +1,35 @@
+const mapStateToProps = (state) => (
+  {
+    items: state.basket.items,
+    totalItems: () => (
```
на взгляд: не должно быть функцией? можно сразу посчитать

index01d commented on Week-1/src/components/views/Checkout/index.js:
```diff
@@ -1,35 +1,44 @@
+
+const mapStateToProps = (state) => (
+  { items: state.basket.items,
+    totalCost: () => {
```
не должно быть функцией, тут уже точно нужно выделить код connect, тк уже есть сложные вычисления не относящиеся к презентационной логике. Условно будем считать, что в index у нас лежит контейнерный компонент, созданный с помощью connect()

violentr replied:
> не должно быть функцией, тут уже точно нужно выделить код connect, тк уже есть сложные вычисления не относящиеся к презентационной логике. Условно будем считать, что в index у нас лежит контейнерный компонент, созданный с помощью connect()

@index01d не очень понятен комментарий, что здесь не верно ? нужно сделать презентационный компонент ? и index.js сделать connect кодом ? и почему totalCost не может быть функцией ?

index01d replied:
конкретно здесь имеется ввиду, что нет смысла totalCost делать функцией:

```
totalCost: state.basket.items.reduce((total, product) => total + product.price, 0)
```

по факту решает задачу

index01d commented on Week-1/src/components/views/Product/index.js:
```diff
@@ -1,26 +1,30 @@
+const mapStateToProps = (state) => (
+  {
+    items: state.products.entries,
+    findById: (products, id) => (
```
получилось длинная цепочка вызовов с пробрасываемой зависимостью (products), выглядит так, что как минимум первый аргумент не нужен

index01d replied:
и второй тоже, ведь второй аргумент mapStateToProps = (state, ownProps) и там, в ownProps как раз искомый id, получается можно сразу передать компоненту нужный продукт

index01d commented on Week-1/src/components/views/Checkout/index.js:
```diff
@@ -1,35 +1,44 @@
+  }
+  render(){
+    return (
+      <Fragment>
```
зачем два вложенных фрагмента?

index01d commented on Week-1/src/components/views/Checkout/index.js:
```diff
@@ -1,35 +1,44 @@
-import React, { Fragment } from 'react'
+import React, { Fragment, Component } from 'react'
 import ProductCard from '~/src/components/Product/Card.js'
 import { imageAttributes } from '~/src/shared/helper.js'
```
выглядит чуть сомнительно, список многовато знает о настройках зависимостей карточки;  нужно или надстраивать CatalogCard или вообще вхардкодить в ProductCard, если аттрибуты одинаковы для всех.

index01d commented on Week-1/src/components/views/Checkout/index.js:
```diff
@@ -1,35 +1,44 @@
+      <Fragment>
+        <Fragment>
+          <h3> This is Checkout page </h3>
+          { listProducts(this.props) }
```
выделение listProducts теперь выглядит избыточным, весь этот код вполне симпатично смотрелся бы и здесь

index01d commented on Week-1/src/components/views/Checkout/index.js:
```diff
@@ -1,35 +1,44 @@
-)
 
-export default CheckoutPage
+class CheckoutPage extends Component {
```
на взгляд компонент мог бы быть обычным функциональным `({ items, totalCost }) => (<..someJSX>)`

violentr committed:
Move router logic from Layout to App component

Router logic should be in one place not to
be as part of Layout, better code semantics

violentr committed:
Add ProductsPage connect to the Redux store

ProductsPage does the role of container in
this case and the rest what depend on it
just Presentation components, no need them
to connect to Redux store

violentr committed:
Add totalItems, no need to make it as function

violentr committed:
Remove totalCost from props, made as Component func

No need to have this logic available through the
props make it as regular component func

violentr committed:
Add Hardcode imageAttributes for ProductCard

violentr committed:
Add Make use of ownProps to get productId

there are 2 params mapStateToProps, state
and ownProps, we use ownProps to get productId

violentr committed:
Add totalCost is not function, better code semantics

index01d commented on Week-1/src/components/Product/Basket.js:
```diff
@@ -1,14 +1,33 @@
+    return (
+      <Fragment>
+        <Link to={checkoutPath()}>
+          <button disabled={ !this.props.totalItems > 0}>
```
== 0? условие какое-то неправильное

violentr replied:
> == 0? условие какое-то неправильное

согласен, че то я здесь намудрил ;-)

index01d commented on Week-1/src/components/views/Products/Catalog.js:
```diff
@@ -1,24 +1,22 @@
+          <ProductCard key={product.id}
+            product={product}
+          />
+          <ProductAddToBasket key={i} item={product} addProduct={addProduct} />
```
каталог теперь слишком много знает о логике корзины. ProductAddToBasket можно же снабдить собственным контейнером и передавать здесь в свойствах только сам product (к тому же стоит свойство назвать product={product}, чтобы не изобретать разных имен одного и того же и не вводить коллег в заблуждение:))

index01d replied:
тогда ProductAddToBasket вполне может себе позволить быть частью ProductCard

violentr replied:
> тогда ProductAddToBasket вполне может себе позволить быть частью ProductCard

@index01d если я сделаю  ProductAddToBasket  частью компонента ProductCard то у меня AddToBasket button будет везде  и в Checkout page
а мне это не надо ;-)

index01d commented on Week-1/src/components/views/Products/index.js:
```diff
@@ -1,19 +1,30 @@
+import {connect} from 'react-redux'
+import {addProductToBasket} from '~/src/actions/Basket.js'
 
 class ProductsPage extends Component {
```
выглядит теперь лишним компонентом, который ничего не делает, можно напрямую здесь приконнектить 
`export default connect(mapStateToProps, actionToProps)(Catalog)` и будет ок

violentr committed:
Add ProductAddToBasket behaves as container

Catalog had a lot information about
AddToBasket functionality what is not
its concern, AddToBasket behaves as
container component

violentr committed:
Remove ProductsPage no need, Catalog used instead

violentr committed:
Fix Routes after ProductsPage removed, used Catalog

violentr committed:
Add Better condition for showing Basket button

violentr committed:
Fix typo actionsToProps
